types/tls.go

```diff
@@ -0,0 +1,188 @@
+package types
+
+import (
+	"fmt"
+)
+
+// TLSVersion is a strongly-typed int used for relative comparison
+// (minimum, maximum, greater than, less than) of TLS versions
+type TLSVersion int
+
+const (
+	// Error value, excluded from lookup maps
+	TLSVersionInvalid TLSVersion = iota - 1
+
+	// Explicit unspecified zero-value to avoid overwriting parent defaults
+	TLSVersionUnspecified
+
+	// Explictly allow implementation to select TLS version
+	// May be useful to supercede defaults specified at a higher layer
+	TLSVersionAuto
+
+	_ // Placeholder for SSLv3, hopefully we won't have to add this
+
+	// TLS versions
+	TLSv1_0
+	TLSv1_1
+	TLSv1_2
+	TLSv1_3
+)
+
+var (
+	TLSVersions = map[string]TLSVersion{
+		"TLS_AUTO": TLSVersionAuto,
+		"TLSv1_0":  TLSv1_0,
+		"TLSv1_1":  TLSv1_1,
+		"TLSv1_2":  TLSv1_2,
+		"TLSv1_3":  TLSv1_3,
+	}
+	// NOTE: This interface is deprecated in favor of TLSVersions
+	// and should be eventually removed in a future release.
+	DeprecatedConsulAgentTLSVersions = map[string]TLSVersion{
+		"":      TLSVersionAuto,
+		"tls10": TLSv1_0,
+		"tls11": TLSv1_1,
+		"tls12": TLSv1_2,
+		"tls13": TLSv1_3,
+	}
+	HumanTLSVersionStrings = map[TLSVersion]string{
+		TLSVersionAuto: "Allow implementation to select TLS version",
+		TLSv1_0:        "TLS 1.0",
+		TLSv1_1:        "TLS 1.1",
+		TLSv1_2:        "TLS 1.2",
+		TLSv1_3:        "TLS 1.3",
+	}
+	ConsulConfigTLSVersionStrings = func() map[TLSVersion]string {
+		inverted := make(map[TLSVersion]string, len(TLSVersions))
+		for k, v := range TLSVersions {
+			inverted[v] = k
+		}
+		return inverted
+	}()
+	// NOTE: these currently map to the deprecated config strings to support the
+	// deployment pattern of upgrading servers first. This map should eventually
+	// be removed and any lookups updated to use ConsulConfigTLSVersionStrings
+	// with newer config strings instead in a future release.
+	ConsulAutoConfigTLSVersionStrings = map[TLSVersion]string{
+		TLSVersionAuto: "",
+		TLSv1_0:        "tls10",
+		TLSv1_1:        "tls11",
+		TLSv1_2:        "tls12",
+		TLSv1_3:        "tls13",
+	}
+)
+
+func (v TLSVersion) String() string {
+	return HumanTLSVersionStrings[v]
+}
+
+func (v TLSVersion) ConsulConfigString() string {
+	return ConsulConfigTLSVersionStrings[v]
+}
```

Comment:
I think `String()` should be the string representation of the enums, not the human-readable string (which I'm not sure we need since it's understandable as-is).

Instead of using `ConsulConfigTLSVersionStrings` I think it's simpler to add a switch-case for all the enum possibilities like:
```go
func (v TLSVersion) String() string {
	switch v {
	case TLSVersionAuto:
		return "TLS_AUTO"
	case TLSv1_0:
		return "TLSv1_0"      
	case TLSv1_1:
		return "TLSv1_1" 
        // ...so on
	default:
		return "INVALID"
	}
}
```

That way the type is actually immutable and we're not exporting unnecessary variables from `types`